cdc/scheduler/util/table_set.go

```diff
@@ -0,0 +1,137 @@
+// scheduler.
+type TableSet struct {
+	tableIDMap   map[model.TableID]*TableRecord
+	captureIndex map[model.CaptureID]map[model.TableID]*TableRecord
+}
+
+// TableRecord is a record to be inserted into TableSet.
+type TableRecord struct {
+	TableID   model.TableID
+	CaptureID model.CaptureID
+	Status    TableStatus
+}
+
+// TableStatus is a type representing the table's replication status.
+// The meaning of its value can be defined by the user of TableSet.
+type TableStatus int32
+
+// NewTableSet creates a new TableSet.
+func NewTableSet() *TableSet {
+	return &TableSet{
+		tableIDMap:   map[model.TableID]*TableRecord{},
+		captureIndex: map[model.CaptureID]map[model.TableID]*TableRecord{},
+	}
+}
+
+// AddTableRecord inserts a new TableRecord.
+// It returns true if it succeeds. Returns false if there is a duplicate.
+func (s *TableSet) AddTableRecord(record *TableRecord) (successful bool) {
+	if _, ok := s.tableIDMap[record.TableID]; ok {
+		// duplicate tableID
+		return false
+	}
+	s.tableIDMap[record.TableID] = record
+
+	captureIndexEntry := s.captureIndex[record.CaptureID]
+	if captureIndexEntry == nil {
+		captureIndexEntry = make(map[model.TableID]*TableRecord)
+		s.captureIndex[record.CaptureID] = captureIndexEntry
+	}
+
+	captureIndexEntry[record.TableID] = record
+	return true
+}
+
+// GetTableRecord tries to obtain a record with the specified tableID.
+func (s *TableSet) GetTableRecord(tableID model.TableID) (*TableRecord, bool) {
+	ret, ok := s.tableIDMap[tableID]
+	return ret, ok
+}
+
+// RemoveTableRecord removes the record with tableID. Returns false
+// if none exists.
+func (s *TableSet) RemoveTableRecord(tableID model.TableID) bool {
+	record, ok := s.tableIDMap[tableID]
+	if !ok {
+		return false
+	}
+	delete(s.tableIDMap, record.TableID)
+
+	captureIndexEntry, ok := s.captureIndex[record.CaptureID]
+	if !ok {
+		panic("unreachable")
+	}
+	delete(captureIndexEntry, record.TableID)
+	if len(captureIndexEntry) == 0 {
+		delete(s.captureIndex, record.CaptureID)
+	}
+	return true
+}
+
+// RemoveTableRecordByCaptureID removes all table records associated with
+// captureID.
+func (s *TableSet) RemoveTableRecordByCaptureID(captureID model.CaptureID) {
+	captureIndexEntry, ok := s.captureIndex[captureID]
+	if !ok {
+		return
+	}
+
+	for tableID := range captureIndexEntry {
+		delete(s.tableIDMap, tableID)
+	}
+	delete(s.captureIndex, captureID)
+}
+
+// CountTableByCaptureID counts the number of tables associated with the captureID.
+func (s *TableSet) CountTableByCaptureID(captureID model.CaptureID) int {
+	return len(s.captureIndex[captureID])
+}
+
+// GetDistinctCaptures counts distinct captures with tables.
+func (s *TableSet) GetDistinctCaptures() []model.CaptureID {
+	var ret []model.CaptureID
+	for captureID := range s.captureIndex {
+		ret = append(ret, captureID)
+	}
+	return ret
+}
+
+// GetAllTables returns all stored information on all tables.
+func (s *TableSet) GetAllTables() map[model.TableID]*TableRecord {
```

Comment:
I am curious about `the definition of table` in this context, it may be `table = tableID + tableStatus`. 

It looks that `GetAllTables` can return a slice of `TableRecord` directly.

Reply:
It would be difficult to look up a slice of `TableRecord`.